Assignment4/haolin.chen-E94054014/task07.py

```diff
@@ -0,0 +1,150 @@
+# -*- coding: utf-8 -*-
+"""Task07.ipynb
+
+Automatically generated by Colaboratory.
+
+Original file is located at
+    https://colab.research.google.com/drive/1RsfoMjRAxYt2kEIRa3sJFg3ugbbg_qQJ
+
+**Task 07: Querying RDF(s)**
+"""
+
+!pip install rdflib
+github_storage = "https://raw.githubusercontent.com/FacultadInformatica-LinkedData/Curso2023-2024/master/Assignment4/course_materials"
+
+"""First let's read the RDF file"""
+
+from rdflib import Graph, Namespace, Literal
+from rdflib.namespace import RDF, RDFS
+g = Graph()
+g.namespace_manager.bind('ns', Namespace("http://somewhere#"), override=False)
+g.namespace_manager.bind('vcard', Namespace("http://www.w3.org/2001/vcard-rdf/3.0#"), override=False)
+g.parse(github_storage+"/rdf/example6.rdf", format="xml")
+
+"""**TASK 7.1: List all subclasses of "LivingThing" with RDFLib and SPARQL**"""
+
+# TO DO
+from rdflib.plugins.sparql import prepareQuery
+
+# Using RDFLib
+print("Using RDFLib:")
+ns = Namespace("http://somewhere#")
+for s, p, o in g.triples((None, RDFS.subClassOf, ns.LivingThing)):
+    print(s)
+
+# Using SPARQL
+print("Using SPARQL:")
+q1 = prepareQuery('''
+    SELECT ?subClass
+    WHERE {
+        ?subClass rdfs:subClassOf ns:LivingThing .
+    }
+    ''',
+    initNs={"rdfs": RDFS, "ns": ns}
+)
+for r in g.query(q1):
+    print(r.subClass)
+
+"""**TASK 7.2: List all individuals of "Person" with RDFLib and SPARQL (remember the subClasses)**
+
+"""
+
+# TO DO
+
+# Using RDFLib
+print("Using RDFLib:")
+ns = Namespace("http://somewhere#")
+for s, p, o in g.triples((None, RDFS.subClassOf, ns.Person)):
+    print(s)
+
+for sub_class, p, o in g.triples((None, RDFS.subClassOf, ns.Person)):
+    for s, p, o in g.triples((None, RDF.type, sub_class)):
+        print(s)
+
+# Using SPARQL
+print("Using SPARQL:")
+q2 = prepareQuery('''
+    SELECT DISTINCT ?individual
+    WHERE {
+        { ?individual rdf:type ns:Person . }
+        UNION
+        { ?subClass rdfs:subClassOf ns:Person .
+          ?individual rdf:type ?subClass . }
+    }
```

Comment:
same issue as above. You can also chain properties with `a/subclassOf*`

Reply:
# update TASK 7.2
q2 = prepareQuery('''
    SELECT ?individual
    WHERE {
        ?individual rdf:type ?type .
        ?type rdfs:subClassOf* ns:Person .
    }
'''